### PR TITLE
Turn rpm-ostree code into a client

### PR DESCRIPTION
**Pull request by ashcrow**
This change allows for modification of rpm-ostree related code so
unittesting can occur. Instead of the functionality being spread
through private functions within the daemon package, a structure
which implements DeploymentClient is used and is passed in to
the Daemon at creation.

Part of https://github.com/openshift/machine-config-operator/issues/101

/cc @jlebon @sdemos @kikisdeliveryservice

### CODE REVIEW

**Comment by ashcrow** [comment]
/hold

I'd like to wait on merging things like this until after the current round of testing.

**Comment by kikisdeliveryservice** [comment]
I'm really excited about these PRs (and future unit testing). Thanks for getting this started!

**Comment by ashcrow** [comment]
Deployment is also an `rpm-ostree` component. I fear there will be more collisions in names like this. I am, however, good with prefixing with something like `NodeDeploymentClient` or `HostDeploymentClient`.

**Comment by jlebon** [comment]
Hmm yeah, I think that makes sense overall for the goal of unit testing. That said for upstream CI, hopefully we can get a real cluster going and test there?

> I am, however, good with prefixing with something like `NodeDeploymentClient` or `HostDeploymentClient`.

Or maybe something like `NodeUpdater`?

**Comment by ashcrow** [comment]
> Or maybe something like NodeUpdater?

I'm :+1: to that. I'll make the change in a bit.

**Comment by ashcrow** [comment]
> Hmm yeah, I think that makes sense overall for the goal of unit testing. That said for upstream CI, hopefully we can get a real cluster going and test there?

Agreed, but that's a different set of work :-)

**Comment by ashcrow** [comment]
Updated.

File: pkg/daemon/rpm-ostree.go [hunk 1]
```diff
@@ -28,7 +28,25 @@ type RpmOstreeDeployment struct {
...
+// This structure implements DeploymentClient
+type RpmOstreeClient struct{}
+
+// NewNodeUpdaterClient returns a new instance of the default DeploymentClient (RpmOstreeClient)
+func NewNodeUpdaterClient() *RpmOstreeClient {
+	return &RpmOstreeClient{}
+}
+
+// GetBootedDeployment returns the current deployment found
+func (r *RpmOstreeClient) GetBootedDeployment(rootMount string) (*RpmOstreeDeployment, error) {
```
Comment (sdemos) [inline pkg/daemon/rpm-ostree.go]: is this function called by anything external to this package? I only see the one change. if not, I would probably advocate for removing it from the client interface and just having it as a method on `RpmOstreeClient`.

Comment (ashcrow) [reply pkg/daemon/rpm-ostree.go]: /cc @jlebon

Comment (ashcrow) [reply pkg/daemon/rpm-ostree.go]: Updated.

**Comment by jlebon** [comment]
Code looks sane to me!

> Should test implementations be in the same file as the main implementation or should they be implemented within a $NAME_test.go with the understanding other $NAME_test.gos will use them as well?

I'll let folks more experienced with golang decide.

**Comment by ashcrow** [comment]
In terms of testing usage. This would look something like this (untested, uncompiled, but close :-)):

```golang
testClient := RpmOstreeClientStub{
	GetBootedOSImageURLReturns: []GetBootedOSImageURLReturn{
            OsImageURL: "theurl",
            Version: "some-version",
            Error: nil,
        }
	RunPivotReturns: []error{nil, fmt.Errorf("broken")}
}

// Create an instance that uses the client
// Execute a method that calls GetBootedOSImageURL off the client. It will get the contents past to GetBootedOSImageURLReturns every time it is executed.
// Execute a method that calls RunPivot. The first time it will return no error. The second run and beyond will return an the "broken" error.

```

File: pkg/daemon/update_test.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+		GetBootedOSImageURLReturns: []GetBootedOSImageURLReturn{},
+		RunPivotReturns: []error{
+			// First run will return no error
+			nil,
+			// Second rrun will return our expected error
+			expectedError},
+	}
+
+	// Create a Daemon instance with mocked clients
+	d := Daemon{
```
Comment (ashcrow) [inline pkg/daemon/update_test.go]: I would expect creating a test `Daemon` instance would likely be added as well to keep many of the tests from recreating the instances each time.

**Comment by ashcrow** [comment]
Moved test code into `*_test.go` and added an example test for `update.go`.

**Comment by ashcrow** [comment]
@sdemos PTAL

**Comment by ashcrow** [comment]
FWIW
```
Running tool: /usr/bin/go test -timeout 30s github.com/openshift/machine-config-operator/pkg/daemon -run ^TestUpdateOS$

ok  	github.com/openshift/machine-config-operator/pkg/daemon	0.007s
Success: Tests passed.
```

**Comment by sdemos** [comment]
just kidding, he is out today, I'll just do it. 

/lgtm

**Pull request merged**